src/librustc/infer/error_reporting/mod.rs

```diff
@@ -1095,15 +1095,16 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                             let sp = hir.span(id);
                             // `sp` only covers `T`, change it so that it covers
                             // `T:` when appropriate
-                            let sp = if has_bounds {
+                            let is_impl_trait = format!("{}", bound_kind).starts_with("impl ");
```

Comment:
`bound_kind.to_string()` is faster; this is error reporting, though, so it's just a nit

Reply:
* `format!()` -> calls https://doc.rust-lang.org/1.31.0/src/alloc/fmt.rs.html#566-573
* `ToString for T: Display` https://doc.rust-lang.org/1.31.0/src/alloc/string.rs.html#2145-2155

They look almost same except that `format!()` uses preallocation. If `to_string` is faster than `format!()`, the preallocation should be removed...

Reply:
Yes, `to_string` is faster by around 37%. The `format!` macro is a complex business and there is definitely room for improvement there, especially in the `Debug` bits.

Reply:
Thanks. 37\% speedup must be due to `ToString`'s specialization for `str`/`String` to avoid using formatting mechanisms, so interesting part is `to_string` is faster (according to #52767) even for other types...